examples/iso_dh/DY.Example.DH.Debug.TraceInvariants.fst

```diff
@@ -0,0 +1,110 @@
+    let (_, tr) = install_public_key alice alice_global_session_pub_key_id (Verify "DH.SigningKey") bob pub_key_bob tr in
+
+    // Store Alice's public key in Bob's state
+    let (priv_key_alice, tr) = get_private_key alice alice_global_session_priv_key_id (Sign "DH.SigningKey") tr in
+    match priv_key_alice with
+    | None -> "Failed to execute protocol run\n"
+    | Some priv_key_alice -> (
+      let pub_key_alice = vk priv_key_alice in
+      let (bob_global_session_pub_key_id, tr) = initialize_pki bob tr in
+      let (_, tr) = install_public_key bob bob_global_session_pub_key_id (Verify "DH.SigningKey") alice pub_key_alice tr in
+
+      let alice_global_session_ids: dh_global_sess_ids = {pki=alice_global_session_pub_key_id; private_keys=alice_global_session_priv_key_id} in
+      let bob_global_session_ids: dh_global_sess_ids = {pki=bob_global_session_pub_key_id; private_keys=bob_global_session_priv_key_id} in
+
+      (*** Run the protocol ***)
+      // Alice
+      prepare_msg1_proof tr alice bob;
+      let (alice_session_id, tr) = prepare_msg1 alice bob tr in
+      assert(trace_invariant tr);
+      
+      send_msg1_proof tr alice bob alice_session_id;
+      let (msg1_id, tr) = send_msg1 alice alice_session_id tr in
+      assert(trace_invariant tr);
+      
+      match msg1_id with
+      | None -> "Failed to execute protocol run\n"
+      | Some msg1_id -> (
+        // Bob
+        prepare_msg2_proof tr alice bob msg1_id;
+        let (bob_session_id, tr) = prepare_msg2 alice bob msg1_id tr in
+        assert(trace_invariant tr);
+        
+        match bob_session_id with
+        | None -> "Failed to execute protocol run\n"
+        | Some bob_session_id -> (
+          send_msg2_proof tr bob_global_session_ids bob bob_session_id;         
+          let (msg2_id, tr) = send_msg2 bob_global_session_ids bob bob_session_id tr in
+          assert(trace_invariant tr);
+          
+          match msg2_id with
+          | None -> "Failed to execute protocol run\n"
+          | Some msg2_id -> (
+            // Alice
+            prepare_msg3_proof tr alice_global_session_ids alice bob msg2_id alice_session_id;
+            let (_, tr) = prepare_msg3 alice_global_session_ids alice bob msg2_id alice_session_id tr in
+            assert(trace_invariant tr);
+            
+            send_msg3_proof tr alice_global_session_ids alice bob alice_session_id;
+            let (msg3_id, tr) = send_msg3 alice_global_session_ids alice bob alice_session_id tr in
+            assert(trace_invariant tr);
+            
+            match msg3_id with
+            | None -> "Failed to execute protocol run\n"
+            | Some msg3_id -> (
+              // Bob
+              verify_msg3_proof tr bob_global_session_ids alice bob msg3_id bob_session_id;
+              let (_, tr) = verify_msg3 bob_global_session_ids alice bob msg3_id bob_session_id tr in
+              assert(trace_invariant tr);
```

Comment:
Looks good to me.

I wonder if some effort could be saved by doing something similar to the model, using just the one debug function and then having a separate external proof that follows it?
Not sure how much effort that would save, but it might be worth looking at.

Reply:
Yes, I wonder what is the technical reason to have duplicate the code testing the protocol?

Reply:
@TWal, do you like the approach with the SMTPats better?

Reply:
I was also quite surprised that it worked, but I would expect that yes, this comes from the encoding of `traceful`.

Does the forward-applied SMTPat do any better or worse for computational load?

Reply:
My guess is that it's roughly the same (one lemma instantiation per monadic function)